### DISCUSSION TITLE
WPF - HwndSource Hook for a child control with its own handle?

### DISCUSSION THREAD

**Discussion by RoyalcraftDev** [dotnet/wpf#9126] [Q&A] [answered]
Hello, I'd like to know if it's possible to use `HwndSource.AddHook` on a custom child control that has its own handle. 

For a little bit of context, I have a control which creates a window using `CreateWindowEx` that is later on rendered in my WPF Window. The issue is that I can't find a way for it to actually handle Win32 messages of the native window I created. I've tried to set the WndProc with `SetWindowLong` but it only results in a `ExecutionEngineException`. The exception can also **vary**. Then I tried to create my own HwndSource but gave me an `ArgumentException` telling me that it has to be an element without a parent, in other words, the Window is the only thing that can be used for the `HwndSource`.

So now, is there a way to actually do it similar to how `HwndSource` works but for child elements?

Note: ~If you want to reproduce the issue, simply create a WPF project, DLL import `SetWindowLong` with a `WndProc` delegate and call it using the window's handle.~ (This isn't the right method to reproduce my issue.)

IMO, the reason of this behavior is a conflict with the WPF way of handling window messages but I'm no expert on the subject, so I might be wrong.

Update: It seems like I am able to set the `WndProc` (I tried to set it after initializing the renderer for the surface.)

**It crashes because my WndProc is triggered. I tried to remove the code and return `nint.Zero`, removing the code and returning `DefWindowProc` but those still trigger the `ExecutionEngineException`**

I set the `WndProc` this way:
![Set the WndProc](https://github.com/dotnet/wpf/assets/61633292/ae7049a5-93b9-4538-acbe-5cfed2032ea2)

Define it like this:
![WndProc def](https://github.com/dotnet/wpf/assets/61633292/8080063d-3554-40db-9aae-ba9cf70c7931)

PInvoke:
![SetWindowLong proc](https://github.com/dotnet/wpf/assets/61633292/bfb14330-d8da-4c42-84d7-fc08c5a94467)
![DefWindowProc](https://github.com/dotnet/wpf/assets/61633292/8240479e-3747-4651-a515-dc4b88006fa5)

**Comment by miloush** [comment]
Please attach a repro project so that we can see your pinvokes and what you are trying to subclass.

**Reply by RoyalcraftDev** [reply]
Yeah, never mind what I said, it's wrong. It doesn't seem to actually do the issue I mentioned when I try to reproduce on a new project. So I will try to be more specific, I load a DLL which take cares of the rendering on the surface I'm creating (I do that on a different thread.) It works fine if I `AddHook()` my `WndProc` but does the issue I mentioned above if I try to set the `WndProc` with `SetWindowLong` directly. I'm not entirely issue how can I get a more accurate stacktrace, since I don't get anything except that it occurs somewhere around the WPF window message handling code. As I said the error can vary, when I re-run the code.

**Reply by RoyalcraftDev** [reply]
By never starting the thread, the issue doesn't happen, so there is somewhat a link between both. But this doesn't give me much info about why since the exception doesn't even have a message.

**Reply by RoyalcraftDev** [reply]
![image](https://github.com/dotnet/wpf/assets/61633292/58e14b9c-bf51-4e11-b12d-610464c80224)
![image](https://github.com/dotnet/wpf/assets/61633292/e239b65a-0438-4572-9589-eb4c9d6a5cfa)
![image](https://github.com/dotnet/wpf/assets/61633292/b2a0f969-133f-4d4e-94b2-fef00b76e47c)
This happens only when the renderer is initialized and the `WndProc` set with the code above. I tried the same thing in WinForms, I don't have that issue. Both versions are almost identical, just adapted to WPF way.

**Reply by RoyalcraftDev** [reply]
I updated the original question.

**Reply by miloush** [reply]
Thanks for providing the pinvokes. I don't see anything wrong with them at first glance. Are you able to attach a repro project for debugging? (you can just drag and drop it in the textbox here)

**Reply by RoyalcraftDev** [reply]
Okay it took me quite a while to reproduce it but now it successfully fails on a new project, I'm uploading it right away!

**Reply by RoyalcraftDev** [reply]
[WpfApp2.zip](https://github.com/dotnet/wpf/files/15377303/WpfApp2.zip)

**Reply by RoyalcraftDev** [reply]
Also note, to trigger the issue you need to play a bit with the window, resize, hover, click, etc. With my actual project it's more direct, hovering directly causes the issue but here it's slower for some reasons.

**Reply by miloush** [reply]
These are symptoms of GC being involved. As @NeverMorewd pointed out below, you have to make sure the pointer you hand to the native code stays valid even when GC moves your WndProc in the memory. Your code is just passing simple pointers that get outdated.

**Reply by RoyalcraftDev** [reply]
Alright I will take a look and let you know

**Comment by NeverMorewd** [comment, marked as answer]
You may refer to https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle?view=net-8.0

**Reply by RoyalcraftDev** [reply]
Thanks!
